intro/intro_tutorial.py

```diff
@@ -0,0 +1,327 @@
+df_dev.sample(5, random_state=3)
+
+# %% [markdown]
+# Each row is one comment consisting of text, author, and date values, as well as an integer id for which YouTube video the comment corresponds to.
+# Additionally, since we are looking at the development set, these examples have labels as well- `1` for spam, `0` for ham (not spam).
+
+# %% [markdown]
+# ## STEP 1: Writing Labeling Functions
+#
+# _Labeling functions (LFs)_ are one of the core operators for building and managing training datasets programmatically in Snorkel.
+# The basic idea is simple: **a labeling function is a function that labels some subset of the training dataset**.
+# That is, each labeling function either outputs `SPAM`, `HAM`, or `ABSTAIN`:
+
+# %%
+HAM = 0
+SPAM = 1
+ABSTAIN = -1
+
+# %% [markdown]
+# Labeling functions can be used to represent many heuristic strategies for labeling data.
+# **The key idea is that labeling functions do not need to be perfectly accurate**, as Snorkel will automatically estimate their accuracies and correlations, and then reweight and combine their output labels, leading to high-quality training labels.
+#
+# As a starting example, labeling functions can be based on heuristically **matching keywords**:
+
+# %%
+from snorkel.labeling.lf import labeling_function
+
+@labeling_function()
+def lf_keyword_my(x):
+    """Many spam comments talk about 'my channel', 'my video', etc."""
+    return SPAM if "my" in x.text.lower() else ABSTAIN
+
+
+# %% [markdown]
+# Labeling functions can also use **regular expressions**:
+
+# %%
+import re
+
+@labeling_function()
+def lf_regex_check_out(x):
+    """Spam comments say 'check out my video', 'check it out', etc."""
+    return SPAM if re.search(r"check.*out", x.text, flags=re.I) else ABSTAIN
+
+
+# %% [markdown]
+# Labeling functions can of course flexibly express a range of other **heuristics**:
+
+# %%
+@labeling_function()
+def lf_short_comment(x):
+    """Ham comments are often short, such as 'cool video!'"""
+    return HAM if len(x.text.split()) < 5 else ABSTAIN
+
+
+# %% [markdown]
+# Labeling functions can also use **third-party models**:
+
+# %%
+from textblob import TextBlob
+
+@labeling_function()
+def lf_textblob_polarity(x):
+    """
+    We use a third-party sentiment classification model, TextBlob,
+    combined with the heuristic that ham comments are often positive.
+    """
+    sentiment = TextBlob(x.text).sentiment
+    return HAM if sentiment.polarity > 0.3 else ABSTAIN
+
+
+# %% [markdown]
+# For many more types of labeling functions---including over data modalities beyond text---see the other [tutorials](https://github.com/snorkel-team/snorkel-tutorials) and examples at [snorkel.org](http://snorkel.org).
+# In general the process of developing labeling functions is, like any other development process, an iterative one that takes time- but that, in [many cases](http://snorkel.org), can be orders-of-magnitude faster that hand-labeling training data.
```

Comment:
snorkel.org was already linked to in the previous line (and two other times before this), so you can probably skip this link.